wasi/wasi_test.go

```diff
@@ -0,0 +1,261 @@
+				argvWritten += uint32(len(arg))
+			}
+
+			// Compare them
+			errno = wasiEnv.args_get(hostFunctionCallContext, argsPtr, argvPtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgs, hostFunctionCallContext.Memory.Buffer[argsPtr:argsPtr+uint32(len(expectedArgs))])
+			require.Equal(t, expectedArgv, hostFunctionCallContext.Memory.Buffer[argvPtr:argvPtr+uint32(len(expectedArgv))])
+		})
+	}
+}
+
+func TestArgsSizesGetReturnError(t *testing.T) {
+	dummyArgs := []string{"foo", "bar", "baz"}
+	argsOpt, err := Args(dummyArgs)
+	require.NoError(t, err)
+	wasiEnv := NewEnvironment(argsOpt)
+	hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+	memorySize := uint32(len(hostFunctionCallContext.Memory.Buffer))
+	inMemory := uint32(0)
+
+	tests := []struct {
+		name           string
+		argsCountPtr   uint32
+		argsBufSizePtr uint32
+	}{
+		{
+			name:           "out-of-bound argsCountPtr",
+			argsCountPtr:   memorySize,
+			argsBufSizePtr: inMemory,
+		},
+		{
+			name:           "out-of-bound argsBufSizePtr",
+			argsCountPtr:   inMemory,
+			argsBufSizePtr: memorySize,
+		},
+		{
+			name:           "barely out-of-bound argsCountPtr",
+			argsCountPtr:   memorySize - SIZE_UINT32 + 1,
+			argsBufSizePtr: inMemory,
+		},
+		{
+			name:           "barely out-of-bound argsBufSizePtr",
+			argsCountPtr:   inMemory,
+			argsBufSizePtr: memorySize - SIZE_UINT32 + 1,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			errno := wasiEnv.args_sizes_get(hostFunctionCallContext, tc.argsCountPtr, tc.argsBufSizePtr)
+			require.Equal(t, EINVAL, errno)
+		})
+	}
+}
+
+func TestArgsGetAPIReturnError(t *testing.T) {
+	dummyArgs := []string{"foo", "bar", "baz"}
+	argsOpt, err := Args(dummyArgs)
+	require.NoError(t, err)
+	wasiEnv := NewEnvironment(argsOpt)
+	hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+	memorySize := uint32(len(hostFunctionCallContext.Memory.Buffer))
+	inMemory := uint32(0)
+	argsArray, err := newWASIStringArray(dummyArgs)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name       string
+		argsPtr    uint32
+		argsBufPtr uint32
+	}{
+		{
+			name:       "out-of-bound argsPtr",
+			argsPtr:    memorySize,
+			argsBufPtr: inMemory,
+		},
+		{
+			name:       "out-of-bound argsBufPtr",
+			argsPtr:    inMemory,
+			argsBufPtr: memorySize,
+		},
+		{
+			name:       "barely out-of-bound argsPtr",
+			argsPtr:    memorySize - SIZE_UINT32*uint32(len(argsArray.nullTerminatedValues)) + 1,
+			argsBufPtr: inMemory,
+		},
+		{
+			name:       "barely out-of-bound argsBufPtr",
+			argsPtr:    inMemory,
+			argsBufPtr: memorySize - argsArray.totalBufSize + 1,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			errno := wasiEnv.args_get(hostFunctionCallContext, tc.argsPtr, tc.argsBufPtr)
+			require.Equal(t, EINVAL, errno)
+		})
+	}
+}
+
+func buildMockHostFunctionCallContext() *wasm.HostFunctionCallContext {
```

Comment:
nit: In what sense this is mock?

Reply:
Yeah, this is a real instance of `HostFunctionCallContext`, and `Mock` was not a right name. Fixed it. Thanks!

Incidentally, the reason why I named it that in the first place is because this function directly builds the `HostFunctionCallContext` and does not guarantee invariant that might be guaranteed if it was built with the original method. For example, in the current implementation, the length of the memory is not a multiple of the page size, which is not a problem in this test though.